File: docs/test-cases/oauth-implementation.py
# Test case: oauth-implementation (A07:2025)
import jwt
from flask import Flask, request, redirect, session

app = Flask(__name__)
app.secret_key = "hardcoded_flask_secret"

TRUSTED_BASE = "https://example.com"


@app.route("/oauth/start")
def oauth_start():
    # BUG: no state parameter — CSRF against OAuth callback possible
    redirect_uri = request.args.get("redirect_uri", "")
    return redirect(
        f"https://idp.example.com/auth?redirect_uri={redirect_uri}&client_id=myapp"
    )


@app.route("/oauth/callback")
def oauth_callback():
    # BUG: prefix match — "https://example.com.evil.com/steal" passes this check
    redirect_uri = request.args.get("redirect_uri", "")
    if not redirect_uri.startswith(TRUSTED_BASE):
        return "Bad redirect", 400

    # BUG: no state validation — no CSRF protection
    code = request.args.get("code")
    token = exchange_code(code, redirect_uri)

    # BUG: algorithms=["none"] accepts unsigned tokens; no audience check
    payload = jwt.decode(token, "secret", algorithms=["none", "HS256"])
    session["user_id"] = payload["sub"]
    return redirect(redirect_uri)


def exchange_code(code: str, redirect_uri: str) -> str:
    # BUG: redirect_uri passed verbatim — no exact-match validation
    import requests
    resp = requests.post("https://idp.example.com/token", data={
        "code": code,
        "redirect_uri": redirect_uri,
        "client_secret": "hardcoded_client_secret_xyz",  # BUG: hardcoded
    })
    return resp.json()["access_token"]

File: docs/test-cases/oauth-implementation.rs
// Test case: oauth-implementation (A07:2025)
use jsonwebtoken::{decode, Algorithm, DecodingKey, Validation};
use serde::Deserialize;

const TRUSTED_BASE: &str = "https://app.example.com";
const SIGNING_KEY: &[u8] = b"hardcoded_hs256_secret";

#[derive(Debug, Deserialize)]
struct Claims {
    sub: String,
    exp: usize,
}

fn oauth_start(redirect_uri: &str) -> String {
    // BUG: no `state` or `nonce` generated — CSRF against OAuth callback
    format!(
        "https://idp.example.com/auth?client_id=myapp&redirect_uri={}",
        redirect_uri
    )
}

fn oauth_callback(redirect_uri: &str, token: &str) -> Result<String, String> {
    // BUG: starts_with prefix match — "https://app.example.com.attacker.io" passes
    if !redirect_uri.starts_with(TRUSTED_BASE) {
        return Err("bad redirect".into());
    }

    // BUG: Algorithm::None accepts unsigned tokens — full signature bypass
    let mut validation = Validation::new(Algorithm::None);
    validation.insecure_disable_signature_validation();
    // BUG: no audience set on the validation — wrong-aud tokens accepted
    validation.validate_aud = false;

    let data = decode::<Claims>(
        token,
        &DecodingKey::from_secret(SIGNING_KEY),
        &validation,
    )
    .map_err(|e| e.to_string())?;

    // BUG: no `state` parameter checked against stored session value
    Ok(format!("welcome {}", data.claims.sub))
}
